### PR TITLE
Added a way to get build configuration info from the agent.

### PR DESCRIPTION
**Pull request by Ferroin**
##### Summary

This adds a new option to the `-W` switch called 'buildinfo'. When invoked with this argument, Netdata will print it's version, the
configure options, and a list of optional features and whether they are enabled or not.

This is intended to serve three purposes:

* It allows developers to more quickly get an idea of how Netdata was built when triaging bug reports.
* It provides an easier way to validate changes to the build system that affect optional features during the development cycle.
* It provides an easier way to build CI workflows that validate that building under a given set of constraints results in a feature being enabled or not.

The actual implementation is a bit large but overall exceedingly simple, consisting of a set of preprocessor directives to extract optional feature state information from config.h and then a series of printf() calls to actually report this info (which should end up optimized by smart compilers due to all the arguments being compile-time constants).

The use of `-W` for this was chosen to avoid needing a new option name and also avoid polluting the `-V` output.

The output looks like this (from a local build on my development system):

```
Version: netdata v1.24.0-220-g16165edb8
Configure options:  'PKG_CONFIG_PATH=/usr/local/lib64/pkgconfig'
Features:
    dbengine:                NO
    Native HTTPS:            NO
    Netdata Cloud:           NO
    TLS Host Verification:   YES
Libraries:
    jemalloc:                NO
    JSON-C:                  YES
    libcap:                  YES
    libcrypto:               NO
    libm:                    YES
    LWS:                     NO
    mosquitto:               NO
    tcalloc:                 NO
    zlib:                    YES
Plugins:
    apps:                    YES
    cgroup Network Tracking: YES
    CUPS:                    YES
    EBPF:                    NO
    IPMI:                    NO
    NFACCT:                  YES
    perf:                    YES
    slabinfo:                YES
    Xen:                     NO
    Xen VBD Error Tracking:  NO
Exporters:
    AWS Kinesis:             NO
    GCP PubSub:              NO
    MongoDB:                 NO
    Prometheus Remote Write: YES
```

Sample POSIX shell one-liner to test for a specific feature in the build using this output (in this case checking for Netdata Cloud support):

```
[ "$(./netdata -W buildinfo | grep 'Netdata Cloud' | awk '{print $2}')" = "YES" ]
```

##### Component Name

area/build
area/daemon

##### Test Plan

Testing is done by simply building the agent and validating the output of `netdata -W buildinfo` against the contents of `config.log`.

##### Additional Information

This type of thing has been discussed at least a couple of times before as a potential addition to the agent. The issues with the libJudy bundling code are what finally prompted me to just sit down and actually implement it, as this would have made it much easier to verify for certain that the original PR was buggy in multiple places.

The diffstat here looks huge, but it's really not all that much in the way of code because most of it is `daemon/buildinfo.c`, which unfortunately has to be procedurally iterative due to a number of quirks of the C preprocessor language.

The output is a bit verbose as a result of being designed to be human friendly while still being easily parsed programmatically (to allow for use in CI).

### CODE REVIEW

**Review by prologic — COMMENTED** [review]
LGTM. Deferring to @netdata/agent team to review and approve.

File: daemon/buildinfo.c [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+#define FEAT_DBENGINE "NO"
+#endif
+
+#if HAVE_X509_VERIFY_PARAM_set1_host
+#define FEAT_TLS_HOST_VERIFY "YES"
+#else
+#define FEAT_TLS_HOST_VERIFY "NO"
+#endif
+
+#ifdef ENABLE_HTPS
```
Comment (stelfrag) [inline daemon/buildinfo.c:26]: ENABLE_HTTPS ?

File: daemon/buildinfo.c [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+#ifdef ENABLE_PROMETHEUS_REMOTE_WRITE
+#define FEAT_REMOTE_WRITE "YES"
+#else
+#define FEAT_REMOTE_WRITE "NO"
+#endif
+
+
+void print_build_info() {
+    int i;
```
Comment (thiagoftsm) [inline daemon/buildinfo.c:178]: This variable is not used, you can remove it.

File: daemon/buildinfo.c (hunk 1 above)
Comment (Ferroin) [reply daemon/buildinfo.c:26]: It's set in `config.h` by the configure script if native HTTPS support is enabled.

File: daemon/buildinfo.c (hunk 2 above)
Comment (Ferroin) [reply daemon/buildinfo.c:178]: Nice catch, not sure how I missed that when I quit trying to do this as looping over an array of options.

File: daemon/buildinfo.c (hunk 1 above)
Comment (Ferroin) [reply daemon/buildinfo.c:26]: To clarify a bit further, all the `#ifdef` checks here are looking for things set in `config.h` by the configure script if certain features are enabled or specific libraries are going to be linked in. I tried a couple of different approaches when initially writing this, but this one looked the least ugly by a pretty significant margin.

Comment (mfundul) [reply daemon/buildinfo.c:26]: Stelios is probably talking about the spelling.

Comment (Ferroin) [reply daemon/buildinfo.c:26]: Ah, yeah, missed that somehow...

File: daemon/buildinfo.c [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+#define FEAT_CLOUD "NO"
+#endif
+
+#ifdef ENABLE_DBENGINE
+#define FEAT_DBENGINE "YES"
+#else
+#define FEAT_DBENGINE "NO"
+#endif
+
+#if HAVE_X509_VERIFY_PARAM_set1_host
```
Comment (mfundul) [inline daemon/buildinfo.c:20]:
```suggestion
#if defined(HAVE_X509_VERIFY_PARAM_set1_host) && HAVE_X509_VERIFY_PARAM_set1_host == 1
```

File: daemon/buildinfo.c [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+#define FEAT_ZLIB "NO"
+#endif
+
+#ifdef STORAGE_WITH_MATH
+#define FEAT_LIBM "YES"
+#else
+#define FEAT_LIBM "NO"
+#endif
+
+#ifdef HAVE_CRYPTO
```
Comment (mfundul) [inline daemon/buildinfo.c:82]: `HAVE_CRYPTO` is not defined anywhere.

File: daemon/buildinfo.c [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+#endif
+
+#ifdef ENABLE_PROMETHEUS_REMOTE_WRITE
+#define FEAT_REMOTE_WRITE "YES"
+#else
+#define FEAT_REMOTE_WRITE "NO"
+#endif
+
+
+void print_build_info() {
```
Comment (mfundul) [inline daemon/buildinfo.c:177]:
```suggestion
void print_build_info(void) {
```

**Review by thiagoftsm — APPROVED** [review]
After @Ferroin to fix all problems, I observed that the output is correct according the libraries I have installed:

```
Configure options:  '--prefix=/usr' '--sysconfdir=/etc' '--localstatedir=/var' '--libexecdir=/usr/libexec' '--libdir=/usr/lib' '--with-zlib' '--with-math' '--with-user=netdata' '--disable-lto' 'CFLAGS=-O1 -ggdb -Wall -Wextra -Wformat-signedness -fstack-protector-all -DNETDATA_INTERNAL_CHECKS=1 -D_FORTIFY_SOURCE=2 -DNETDATA_VERIFY_LOCKS=1' 'LDFLAGS=' 'PKG_CONFIG_PATH=/usr/local/lib64/pkgconfig:/usr/local/share/pkgconfig:/usr/lib64/pkgconfig:/usr/share/pkgconfig:/usr/local/lib64/pkgconfig:/usr/local/share/pkgconfig'
Features:
    dbengine:                YES
    Native HTTPS:            YES
    Netdata Cloud:           YES
    TLS Host Verification:   YES
Libraries:
    jemalloc:                NO
    JSON-C:                  YES
    libcap:                  YES
    libcrypto:               YES
    libm:                    YES
    LWS:                     YES
    mosquitto:               YES
    tcalloc:                 NO
    zlib:                    YES
Plugins:
    apps:                    YES
    cgroup Network Tracking: YES
    CUPS:                    YES
    EBPF:                    YES
    IPMI:                    YES
    NFACCT:                  YES
    perf:                    YES
    slabinfo:                YES
    Xen:                     NO
    Xen VBD Error Tracking:  NO
Exporters:
    AWS Kinesis:             YES
    GCP PubSub:              NO
    MongoDB:                 YES
    Prometheus Remote Write: YES
```

LGTM!

**Review by mfundul — APPROVED** [review]

**Review by prologic — APPROVED** [review]
Has my approval too since the C devs have also approved 😀

**Pull request merged**